pkg/utils/kubernetes/health/health.go

```diff
@@ -0,0 +1,423 @@
+// SPDX-FileCopyrightText: 2021 SAP SE or an SAP affiliate company and Gardener contributors.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package health
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/go-logr/logr"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	apimacherrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	kutil "github.com/gardener/landscaper/pkg/utils/kubernetes"
+)
+
+const (
+	StatusHealthy    StatusType = "Healthy"
+	StatusNotHealthy StatusType = "NotHealthy"
+	StatusUnknown    StatusType = "Unknown"
+)
+
+// StatusType defines the value of a Status.
+type StatusType string
+
+// WaitForObjectsHealthy waits for objects to be heatlhy and
+// returns an error if all the objects are not healthy after the timeout.
+func WaitForObjectsHealthy(ctx context.Context, timeout time.Duration, log logr.Logger, kubeClient client.Client, objects []*unstructured.Unstructured) error {
+	var (
+		wg  sync.WaitGroup
+		try int32 = 1
+
+		// allErrs contains all the errors not related to the healthiness of objects.
+		allErrs []error
+		// notHealthyErrs contains all the errors related to the healthiness of objects.
+		notHealthyErrs []error
+	)
+
+	_ = wait.PollImmediate(5*time.Second, timeout, func() (bool, error) {
+		log.V(3).Info("wait resources healthy", "try", try)
+		try++
+
+		allErrs = nil
+		notHealthyErrs = nil
+		for _, obj := range objects {
+			wg.Add(1)
+			go func(obj *unstructured.Unstructured) {
+				defer wg.Done()
+
+				if err := IsObjectHealthy(ctx, log, kubeClient, obj); err != nil {
+					switch err.(type) {
+					case *ObjectNotHealthyError:
+						notHealthyErrs = append(notHealthyErrs, err)
+					default:
+						allErrs = append(allErrs, err)
+					}
+				}
+			}(obj)
+		}
+		wg.Wait()
+
+		if len(allErrs) > 0 {
+			return false, apimacherrors.NewAggregate(allErrs)
+		}
+		if len(notHealthyErrs) > 0 {
+			return false, nil
+		}
+
+		return true, nil
+	})
+
+	if len(allErrs) > 0 {
+		return apimacherrors.NewAggregate(allErrs)
+	}
+	if len(notHealthyErrs) > 0 {
+		return apimacherrors.NewAggregate(notHealthyErrs)
+	}
+
+	return nil
+}
+
+// ObjectNotHealthyError holds information about an unhealthy object
+// and implements the go error interface.
+type ObjectNotHealthyError struct {
+	kind      string
```

Comment:
as this filed does not only contain the kind we should rename it to `gvk` or something similar

Reply:
It contains the return value of `*unstructured.Unstructured.GetKind()` which is indeed just the kind, no group, no version. ;-)

Reply:
Done.